### PR TITLE
Fix #266: Extended NoInfer adding extraSymbols field

### LINKED ISSUE #266: ShowSynthetics
It would be nice to have a rewrite that expands all sugars so that inferred type parameters/implicits are made explicit.

```scala
// before
  val x = List(1).map(_ + 1)
  x + "foobar"
// after
  val x = List.apply[Int](1).map[Int, List[Int]](_ + 1)(scala.collection.immutable.List.canBuildFrom[Int])
  scala.Predef.any2stringadd[List[Int]](x) + "foobar"
```

### PR DESCRIPTION
**Pull request by bmatthews68**
Done in Scala Center's spree in Dublin.

### CODE REVIEW

File: scalafix-core/shared/src/main/scala/scalafix/internal/config/NoInferConfig.scala [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import org.langmeta._
+
+case class NoInferConfig(
+    extraSymbols: List[Symbol] = Nil
+) {
+  implicit val reader: ConfDecoder[NoInferConfig] =
+    ConfDecoder.instanceF[NoInferConfig] { c =>
+        c.getOrElse("extraSymbols")(extraSymbols).map(xs => NoInferConfig(xs))
+    }
+  implicit val symbolReader: ConfDecoder[Symbol] =
```
Comment (olafurpg) [inline scalafix-core/shared/src/main/scala/scalafix/internal/config/NoInferConfig.scala]: There is a Symbol.Global reader in `scalafix.internal.config` which we should use instead, it handles a few corner cases.

File: scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala [hunk 2]
```diff
@@ -1,25 +1,39 @@
...
     with Product {
-  private val badSymbol = SymbolMatcher.exact(NoInfer.badSymbols: _*)
+  private val badSymbol = SymbolMatcher.exact(NoInfer.badSymbols ++ config.extraSymbols: _*)
+
+  def this(index: SemanticdbIndex) =
+    this(index, NoInferConfig.default)
+
+  override def init(config: Conf): Configured[Rule] =
+    config
+      .getOrElse("noInfer")(NoInferConfig.default)
```
Comment (olafurpg) [inline scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala]: I'm starting to think a good convention is to make the configuration key match exactly with the rule name (instead of lowercasing first character like I've done until now). What do you think?

File: scalafix-sbt/src/sbt-test/sbt-scalafix/cross-build/.scalafix.conf [hunk 3]
```diff
@@ -1,4 +1,8 @@
 rules = [
   ProcedureSyntax
   NoAutoTupling
-]
\ No newline at end of file
+]
+
+NoInfer.extraSymbols = [
```
Comment (olafurpg) [inline scalafix-sbt/src/sbt-test/sbt-scalafix/cross-build/.scalafix.conf]: Is this change necessary?

File: scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala [hunk 4]
```diff
@@ -1,25 +1,39 @@
...
     LintCategory.error(
       """The Scala compiler sometimes infers a too generic type such as Any.
         |If this is intended behavior, then the type should be explicitly type
         |annotated in the source.""".stripMargin
     )
-  override def check(ctx: RuleCtx): Seq[LintMessage] =
-    ctx.index.synthetics.flatMap {
+
+  override def check(ctx: RuleCtx): Seq[LintMessage] = {
+
```
Comment (olafurpg) [inline scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala]: Is this change necessary?

File: scalafix-tests/input/src/main/scala/test/NoInfer.scala [hunk 5]
```diff
@@ -14,4 +17,6 @@ case object NoInfer {
   }
   case class A()
   List(NoInfer, A())// assert: NoInfer.product
+  case class B()
+  List(B())// assert: NoInfer.b
```
Comment (olafurpg) [inline scalafix-tests/input/src/main/scala/test/NoInfer.scala]: Nice! :+1:

File: scalafix-core/shared/src/main/scala/scalafix/internal/config/NoInferConfig.scala [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package scalafix.internal.config
+
+import metaconfig._
+import org.langmeta._
+
+case class NoInferConfig(
```
Comment (olafurpg) [inline scalafix-core/shared/src/main/scala/scalafix/internal/config/NoInferConfig.scala]: God there is an incredible amount of ceremony involved around adding configuration, we should try to come up with something easier 🤔

File: scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala (hunk 2 above)
Comment (gabro) [reply scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala]: 👍 to this

File: scalafix-sbt/src/sbt-test/sbt-scalafix/cross-build/.scalafix.conf (hunk 3 above)
Comment (bmatthews68) [reply scalafix-sbt/src/sbt-test/sbt-scalafix/cross-build/.scalafix.conf]: No. I will take it out.

File: scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala (hunk 4 above)
Comment (bmatthews68) [reply scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala]: No. I didn't undo enough when backtracking a change.

File: scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala (hunk 2 above)
Comment (bmatthews68) [reply scalafix-core/shared/src/main/scala/scalafix/internal/rule/NoInfer.scala]: I agree. It would feel more intuitive without the lower case.

**Comment by olafurpg** [comment]
Superseded by #398

**Comment by olafurpg** [comment]
Thank you for this PR @bmatthews68 !
